src/main/java/org/folio/oaipmh/helpers/CQLQueryBuilder.java

```diff
@@ -30,15 +22,46 @@ public CQLQueryBuilder(boolean addQuestionMark) {
     prefix = addQuestionMark ? "?query=" : "&query=";
   }
 
+  /**
+   * Adds a statement to search by source to the query.
+   *
+   * @param source the source to search by
+   * @return {@link CQLQueryBuilder}
+   */
   public CQLQueryBuilder source(String source) {
-    builder.append(String.format(SOURCE_STATEMENT, source));
+    builder.append(String.format("source==%s", source));
     return this;
   }
 
+  /**
+   * Adds a statement to search by id to the query.
+   *
+   * @param id the id to search by
+   * @return {@link CQLQueryBuilder}
+   */
+  public CQLQueryBuilder identifier(String id) {
+    builder.append(String.format("id==%s", id));
+    return this;
+  }
+
+  /**
+   * Adds a statement to search by date range (or just lower/upper bound) to the query.
+   *
+   * @param from the lower date bound
+   * @param until the upper date bound
+   * @return {@link CQLQueryBuilder}
+   */
   public CQLQueryBuilder dateRange(String from, String until) {
-    Map<String, String> valueMap = ImmutableMap.of("from", from, "until", until);
-    StrSubstitutor sub = new StrSubstitutor(valueMap, "%{", "}");
-    builder.append(sub.replace(DATE_RANGE_STATEMENT));
+    builder
+      .append("(")
+      .append(String.format("(metadata.updatedDate==null%s%s)",
+        isNotEmpty(from) ? String.format(" and metadata.createdDate>%s", from) : EMPTY,
```

Comment:
I've just checked specification in more details and see following in `2.7.1 Selective Harvesting and Datestamps` section:
>Range limits are inclusive: from specifies a bound that must be interpreted as "greater than or equal to", until specifies a bound that must be interpreted as "less than or equal to". Therefore, the from argument must be less than or equal to the until argument. Otherwise, a repository must issue a badArgument error.

Could you please update conditions and also validation?

Thanks in advance!

Reply:
Well spotted. Fixed.
I also updated the validation.